File: src/Behavior/Behavior.cpp
#include "Behavior.hpp"
#include "Logger.hpp"

void Behavior::handleMessage(
   MessagePtr message)
{
   // enable
   if (message->getMessageId() == "enable")
   {
      Logger::logDebug(F("Behavior::handleMessage: %s.enable()"), getId().c_str());

      enable();
      message->setFree();
   }
   // disable
   else if (message->getMessageId() == "disable")
   {
      Logger::logDebug(F("Behavior::handleMessage: %s.disable()"), getId().c_str());

      disable();
      message->setFree();
   }
   else
   {
      Component::handleMessage(message);
   }
}

void Behavior::setup()
{
   Component::setup();

   for (Set<Behavior*>::Iterator it = children.begin(); it != children.end(); it++)
   {
      (*it)->setup();
   }
}

void Behavior::loop()
{
   Component::loop();

   if (isEnabled())
   {
      for (Set<Behavior*>::Iterator it = children.begin(); it != children.end(); it++)
      {
         (*it)->loop();
      }
   }
}

void Behavior::setState(
   const int& state)
{
   if (this->state != state)
   {
      int previousState = this->state;
      this->state = state;

      for (Set<BehaviorListener*>::Iterator it = listeners.begin(); it != listeners.end(); it++)
      {
         (*it)->onStateChange(this, previousState, state);
      }
   }
}

File: robox.js/component/behavior.js
function Behavior(id)
{
   this.id = id;
   
   this.parent = null;
   
   Behavior.prototype.ping = function()
   {
      var message = new Message();
      message.messageId = "ping";
      message.destination = this.id;

      parent.sendMessage(message);
   }
   
   Behavior.prototype.enable = function()
   {
      var message = new Message();
      message.messageId = "enable";
      message.destination = this.id;
      
      this.parent.sendMessage(message);
   }
   
   Behavior.prototype.disable = function()
   {
      var message = new Message();
      message.messageId = "disable";
      message.destination = this.id;
      
      this.parent.sendMessage(message);
   }
}